File: backend/schemas/voiceover_plan.py
"""口播生产线 · 分段脚本数据模型（里程碑 A）。"""
from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, List, Optional

from pydantic import BaseModel, Field, field_validator

MAX_VOICEOVER_SEGMENTS = 24
DEFAULT_VOICEOVER_VOICE = "zh-CN-XiaoxiaoNeural"


class VoiceoverPlanStatus(str, Enum):
    DRAFT = "draft"
    CONFIRMED = "confirmed"
    EXECUTING = "executing"
    COMPLETED = "completed"
    FAILED = "failed"


class VoiceoverSegmentStatus(str, Enum):
    DRAFT = "draft"
    SCRIPT_CONFIRMED = "script_confirmed"
    TTS_DONE = "tts_done"
    BROLL_DONE = "broll_done"
    FAILED = "failed"


class VoiceoverWordTiming(BaseModel):
    text: str = ""
    start_sec: float = 0.0
    end_sec: float = 0.0


class VoiceoverTtsState(BaseModel):
    asset_id: Optional[str] = None
    audio_clip_id: Optional[str] = None
    duration_sec: Optional[float] = None
    timeline_start_sec: Optional[float] = None
    word_timings: List[VoiceoverWordTiming] = Field(default_factory=list)


class VoiceoverSubtitleState(BaseModel):
    overlay_ids: List[str] = Field(default_factory=list)
    alignment: Optional[str] = Field(default=None, description="regrouped | sentence | word")


class VoiceoverSearchResult(BaseModel):
    platform: str = ""
    title: str = ""
    url: str = ""
    external_id: Optional[str] = None
    duration_sec: Optional[float] = None
    in_library: bool = False
    library_asset_id: Optional[str] = None


class VoiceoverBrollState(BaseModel):
    search_results: List[VoiceoverSearchResult] = Field(default_factory=list)
    selected: Optional[VoiceoverSearchResult] = None
    library_asset_id: Optional[str] = None
    block_id: Optional[str] = None
    source_in_sec: Optional[float] = None
    source_out_sec: Optional[float] = None
    selection_reason: str = ""


class VoiceoverSegment(BaseModel):
    id: str
    index: int = Field(ge=1, le=MAX_VOICEOVER_SEGMENTS)
    narration_text: str = Field(min_length=1, max_length=4000)
    visual_brief: str = Field(default="", max_length=2000)
    search_queries: List[str] = Field(default_factory=list, max_length=8)
    status: VoiceoverSegmentStatus = VoiceoverSegmentStatus.DRAFT
    tts: VoiceoverTtsState = Field(default_factory=VoiceoverTtsState)
    subtitles: VoiceoverSubtitleState = Field(default_factory=VoiceoverSubtitleState)
    broll: VoiceoverBrollState = Field(default_factory=VoiceoverBrollState)
    error: Optional[str] = None

    @field_validator("search_queries")
    @classmethod
    def normalize_queries(cls, value: List[str]) -> List[str]:
        cleaned: List[str] = []
        for item in value:
            text = str(item or "").strip()
            if text and text not in cleaned:
                cleaned.append(text)
        return cleaned[:8]


class VoiceoverPlan(BaseModel):
    id: str
    status: VoiceoverPlanStatus = VoiceoverPlanStatus.DRAFT
    voice_id: str = DEFAULT_VOICEOVER_VOICE
    speech_rate: str = "+0%"
    user_brief: str = ""
    placeholder_library_asset_id: Optional[str] = None
    segments: List[VoiceoverSegment] = Field(default_factory=list, max_length=MAX_VOICEOVER_SEGMENTS)

    @field_validator("segments")
    @classmethod
    def validate_segment_count(cls, value: List[VoiceoverSegment]) -> List[VoiceoverSegment]:
        if len(value) > MAX_VOICEOVER_SEGMENTS:
            raise ValueError(f"口播分段最多 {MAX_VOICEOVER_SEGMENTS} 段")
        return value


class VoiceoverGenerateRequest(BaseModel):
    user_brief: str = Field(min_length=1, max_length=12000)
    voice_id: Optional[str] = None
    speech_rate: str = "+0%"
    replace_existing: bool = False


class VoiceoverUpdatePlanRequest(BaseModel):
    plan: VoiceoverPlan


class VoiceoverRegenerateSegmentRequest(BaseModel):
    segment_id: str = Field(min_length=1)
    instruction: Optional[str] = Field(default=None, max_length=2000)


class VoiceoverPlanResponse(BaseModel):
    session: "EditSession"
    plan: Optional[VoiceoverPlan] = None


class VoiceoverGenerateResponse(BaseModel):
    session: "EditSession"
    plan: VoiceoverPlan
    note: str = ""


class VoiceoverExecuteRequest(BaseModel):
    placeholder_library_asset_id: Optional[str] = None
    segment_ids: Optional[List[str]] = Field(default=None, max_length=MAX_VOICEOVER_SEGMENTS)


class VoiceoverExecuteResponse(BaseModel):
    session: "EditSession"
    plan: VoiceoverPlan
    note: str = ""


class VoiceoverSearchMaterialsRequest(BaseModel):
    platform: str = Field(default="youtube", description="youtube | bilibili")
    limit: int = Field(default=10, ge=1, le=20)


class VoiceoverSelectMaterialRequest(BaseModel):
    library_asset_id: Optional[str] = None
    search_result: Optional[VoiceoverSearchResult] = None
    search_result_index: Optional[int] = Field(default=None, ge=0, le=19)


class VoiceoverApplyBrollRequest(BaseModel):
    source_in_sec: Optional[float] = Field(default=None, ge=0)
    source_out_sec: Optional[float] = Field(default=None, ge=0)
    wait_download_timeout_sec: float = Field(default=180.0, ge=10.0, le=600.0)


if TYPE_CHECKING:
    from backend.schemas.edit_session import EditSession

File: frontend/src/types/voiceoverPlan.ts
export type VoiceoverPlanStatus =
  | 'draft'
  | 'confirmed'
  | 'executing'
  | 'completed'
  | 'failed'

export type VoiceoverSegmentStatus =
  | 'draft'
  | 'script_confirmed'
  | 'tts_done'
  | 'broll_done'
  | 'failed'

export interface VoiceoverWordTiming {
  text: string
  start_sec: number
  end_sec: number
}

export interface VoiceoverTtsState {
  asset_id?: string | null
  audio_clip_id?: string | null
  duration_sec?: number | null
  timeline_start_sec?: number | null
  word_timings?: VoiceoverWordTiming[]
}

export interface VoiceoverSubtitleState {
  overlay_ids?: string[]
  alignment?: 'sentence' | 'word' | null
}

export interface VoiceoverSearchResult {
  platform: string
  title: string
  url: string
  external_id?: string | null
  duration_sec?: number | null
  in_library?: boolean
  library_asset_id?: string | null
}

export interface VoiceoverBrollState {
  search_results?: VoiceoverSearchResult[]
  selected?: VoiceoverSearchResult | null
  library_asset_id?: string | null
  block_id?: string | null
  source_in_sec?: number | null
  source_out_sec?: number | null
  selection_reason?: string
}

export interface VoiceoverSegment {
  id: string
  index: number
  narration_text: string
  visual_brief: string
  search_queries: string[]
  status: VoiceoverSegmentStatus
  tts?: VoiceoverTtsState
  subtitles?: VoiceoverSubtitleState
  broll?: VoiceoverBrollState
  error?: string | null
}

export interface VoiceoverPlan {
  id: string
  status: VoiceoverPlanStatus
  voice_id: string
  speech_rate: string
  user_brief: string
  placeholder_library_asset_id?: string | null
  segments: VoiceoverSegment[]
}

export interface VoiceoverGenerateRequest {
  user_brief: string
  voice_id?: string
  speech_rate?: string
  replace_existing?: boolean
}

export interface VoiceoverPlanResponse {
  session: import('./editSession').EditSession
  plan?: VoiceoverPlan | null
}

export interface VoiceoverGenerateResponse {
  session: import('./editSession').EditSession
  plan: VoiceoverPlan
  note: string
}

export interface VoiceoverExecuteRequest {
  placeholder_library_asset_id?: string | null
  segment_ids?: string[] | null
}

export interface VoiceoverExecuteResponse {
  session: import('./editSession').EditSession
  plan: VoiceoverPlan
  note: string
}

export interface VoiceoverSearchMaterialsRequest {
  platform?: string
  limit?: number
  search_queries?: string[] | null
}

export interface VoiceoverUpdateSegmentSearchQueriesRequest {
  search_queries: string[]
}

export type VoiceoverSearchQueryLanguage = 'zh' | 'en' | 'ja' | 'ko'

export interface VoiceoverTranslateSearchQueriesRequest {
  target_language: VoiceoverSearchQueryLanguage
  search_queries?: string[] | null
}

export interface VoiceoverSelectMaterialRequest {
  library_asset_id?: string | null
  search_result?: VoiceoverSearchResult | null
  search_result_index?: number | null
}

export interface VoiceoverApplyBrollRequest {
  source_in_sec?: number | null
  source_out_sec?: number | null
  wait_download_timeout_sec?: number
}

export const MAX_VOICEOVER_SEGMENTS = 24

export const VOICEOVER_PLAN_STATUS_LABEL: Record<VoiceoverPlanStatus, string> = {
  draft: '草稿',
  confirmed: '已确认',
  executing: '执行中',
  completed: '已完成',
  failed: '失败',
}

export const VOICEOVER_SEGMENT_STATUS_LABEL: Record<VoiceoverSegmentStatus, string> = {
  draft: '草稿',
  script_confirmed: '脚本已确认',
  tts_done: 'TTS 完成',
  broll_done: '素材完成',
  failed: '失败',
}
